Fix SILCombine of existential applies

meg-gupta opened the pull request:
https://github.com/swiftlang/swift/pull/78354 enabled silcombine to propagate concrete type of existentials in ossa. However when the concrete value is `owned` it is consumed by the `init_existential_ref` forwarding operation. In this case we create an unchecked cast of the opened existential to the concrete type and use that in the apply.

meg-gupta commented:
@swift-ci test

meg-gupta committed:
Remove complex bailouts and fix replacing open_existential_ref

This code was previously creating illegal unchecked_ref_cast
 with same source and destination.

eeckstein commented:
I don't think this is a good idea. Inserting a copy+destroy might be worse than not propagating the concrete type of existential. The optimizer should never insert copies which are not there in the first place.

I suggest to just not do the optimization in such a case.
The correct solution would be to insert a borrow scope around the lifetime of the concrete value and do the optimization on the borrowed value. But this is more complicated and we can do this later - if we see that it is a common case which matters.

eeckstein commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -789,7 +789,12 @@ SILCombiner::buildConcreteOpenedExistentialInfoFromSoleConformingType(
+  // existential.
+  if (SoleCEI.ConcreteValue &&
+      (!F->hasOwnership() ||
+       SoleCEI.ConcreteValue->getOwnershipKind() != OwnershipKind::Owned))
```
Hm, can we still optimize if the `open_existential_ref` is the only use of the `init_existential_ref`? 
Then this pair of instructions can be removed and no copy is required.

meg-gupta replied:
`open_existential_ref` also needs to have a single user i.e apply which will be converted to concrete value for this to work. The optimization also supports introducing unconditional cast for `open_existential_ref` to the concrete type and using the cast value in the apply. I'll try that since we at least get a concrete apply in that case.

meg-gupta committed:
Fix SILCombine of existential applies

Don't use previously found owned concrete values in ossa. They will consumed by
forwarding operations like init_existential_ref. Instead create an unconditional cast
of the opened existential to concrete type and use that to create a concrete apply.

meg-gupta commented:
@swift-ci test

meg-gupta commented on lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp:
```diff
@@ -1304,6 +1301,20 @@ SILInstruction *SILCombiner::createApplyWithConcreteType(
   }
+  // Replace all uses of the opened existential with the unconditional cast to
+  // concrete type.
+  for (auto &valuePair : valuesToReplace) {
```
Changed to MapVector

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test source compatibility

eeckstein reviewed (approved):
lgtm

meg-gupta commented:
@swift-ci smoke test Linux platform

meg-gupta commented:
@swift-ci smoke test Linux platform
